### ISSUE TITLE
Add ability to hook in between build and push steps

### ISSUE THREAD

**Issue by wfg** [GoogleContainerTools/kaniko#1638]
Users want to scan their images after being built by kaniko, but before being pushed to a registry. A practical example is if the image being built does not satisfy their vulnerability policy, the build would fail and not be pushed into the registry. Given that kaniko builds and pushes in one shot, this is not possible today.

I propose adding the ability to hook in between the build and push steps. This would give users flexibility to run scans against their images before pushing them to registries. Perhaps something like a `--post-build` or `--pre-push` flag to which you'd pass a command to run.

This would also address the similar issue https://github.com/GoogleContainerTools/kaniko/issues/563.

**Comment by tejal29** [comment]
Thanks @wfg for the question. 

What you suggest makes complete sense and would be great to add a flag like you suggest to scan the image built. 

Do you if the tool you want to hook with support running a scan in image.tar?

Kaniko does spit out an image tar if you use `--tarPath` flag. If that is the case, you can schedule a container after kaniko to consume the tar binary and run the hook.
e.g.
```


apiVersion: v1
kind: Pod
metadata:
  name: kaniko
spec:
  containers:
  - name: kaniko
    image: gcr.io/kaniko-project/executor:latest
    args: ["--dockerfile=/workspace/dockerfile",
            "--context=dir://workspace",
            "--destination=gcr.io/<repo>/<image>:<tag>" # replace with your image
            "--tarPath=/workspace/image.tar"
            "--no-push"]
   - name: policy-checker   ## Check policy
     image: gcr.io/blah/policy-checker.   ### image that holds the binary for policy checker.
     args: [XXX]

  - name: gcrane 
     image: gcr.io/go-containerregistry/gcrane
     args: ["push", "image.tar", "gcr.io/<repo>/<image>:<tag>"
```

I understand, this is a convoluted way which you can play with right now. 

A couple of questions around the `hooks` for kaniko. How do you plan to bundle the script in the kaniko executor image?
1) Do you extend the current executor image and add a script ? or
2) Copy the script in the kaniko container?

**Comment by wfg** [comment]
@tejal29 I appreciate your time.

---

> Do you if the tool you want to hook with support running a scan in image.tar?

Yes, I'm imagining something like this:
1. kaniko builds the image
2. saves it to a specified location (`--tarPath`)
3. checks for and allows any user-specified commands to run (in our case it would be a scan tool)
4. pushes the image to the registry based on exit code of the user-specified command (if any)

---

I don't think that pod spec would work since it would require the containers to run sequentially (build then scan then push). From what I can tell, that's not possible without `initContainers` instead of `containers`, and even then it would be messy.

---

> A couple of questions around the hooks for kaniko. How do you plan to bundle the script in the kaniko executor image?
> 
> 1. Do you extend the current executor image and add a script ? or
> 2. Copy the script in the kaniko container?

Either of these could work, but the second one sounds better to me. In any case, making the script available for kaniko would be on the user to handle. Kaniko would only need to provide the ability to call the script.

**Comment by tejal29** [comment]
> I don't think that pod spec would work since it would require the containers to run sequentially (build then scan then push). From what I can tell, that's not possible without `initContainers` instead of `containers`, and even then it would be messy.

yes. you are correct. I forgot the fact that all containers run in parallel. It will be messy if you use containers
I will try to take address this feature in next release.

**Comment by briandealwis** [comment]
An alternative might be to use Tekton to perform this kind of orchestration?  Here's [an example](https://ruzickap.github.io/k8s-knative-gitlab-harbor/part-07/) that uses Kaniko with `--tarPath`.

**Comment by vektorich** [comment]
> @tejal29 I appreciate your time.
> 
> > Do you if the tool you want to hook with support running a scan in image.tar?
> 
> Yes, I'm imagining something like this:
> 
> 1. kaniko builds the image
> 2. saves it to a specified location (`--tarPath`)
> 3. checks for and allows any user-specified commands to run (in our case it would be a scan tool)
> 4. pushes the image to the registry based on exit code of the user-specified command (if any)
> 
> I don't think that pod spec would work since it would require the containers to run sequentially (build then scan then push). From what I can tell, that's not possible without `initContainers` instead of `containers`, and even then it would be messy.
> 
> > A couple of questions around the hooks for kaniko. How do you plan to bundle the script in the kaniko executor image?
> > 
> > 1. Do you extend the current executor image and add a script ? or
> > 2. Copy the script in the kaniko container?
> 
> Either of these could work, but the second one sounds better to me. In any case, making the script available for kaniko would be on the user to handle. Kaniko would only need to provide the ability to call the script.

I'm also very interested in this feature
